rfcs/0001-adding-package-logic-to-loader.md

```diff
@@ -0,0 +1,37 @@
+- Feature Name: adding_python_package_logic_to_loader
+- Start Date: 2018-06-29
+- RFC PR:
+- Salt Issue:
+
+# Summary
+[summary]: #summary
+
+Add support for python packages to the loader. Instead of just having
+apache.py, we would have `salt/modules/apache/__init__.py`, and all the other
+modules under that directory loaded in the `apache.<function>` namespace that
+we already have.
+
+# Motivation
+[motivation]: #motivation
+
+Split modules out into different git repositories, and allow the community more control to help maintain them.
+
+# Design
+[design]: #detailed-design
+
+This should be pretty straight forward, we should be able to use the same
+_module_dirs that we are using now, and iterate over the files, and if they are
+a directory, then we load all of the modules underneath it into the namespace
+of the directory.
+
+Heirarchy would be alphabetical, and we would need to log messages if the
+function exists in an earlier file that has been loaded.
+
+## Alternatives
+[alternatives]: #alternatives
```

Comment:
Categorise modules to at least three parts:

- Core modules. Modules, that should be working _everywhere_ (cross-platform) and are vital for a Salt primary usage (package installations, configuration changes).

- All 3rd party modules. Modules that requires additional and optional deps and aren't generic (e.g. Facebook's Cassandra etc).

- All cloud related modules.

This also valid to states and so the states that are related to the 2nd or 3rd category should be moved along with these modules as well.

First layer stays with the Salt's core within the same repo as they are essential. The other two categories should have their own repos.